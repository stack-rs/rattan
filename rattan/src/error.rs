#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("MacParseError: {0}")]
    MacParseError(#[from] MacParseError),
    #[error("NsError: {0}")]
    NsError(#[from] NsError),
    #[error("VethError: {0}")]
    VethError(#[from] VethError),
    #[error("Encounter IO error, {0}")]
    IoError(#[from] std::io::Error),
    #[error("Metal error: {0}")]
    MetalError(#[from] crate::metal::error::MetalError),
    #[error("Tokio Runtime error: {0}")]
    TokioRuntimeError(#[from] TokioRuntimeError),
    #[error("Rattan radix error: {0}")]
    RattanRadixError(String),
    #[error("Rattan core error: {0}")]
    RattanCoreError(#[from] RattanCoreError),
    #[error("Rattan operation error: {0}")]
    RattanOpError(#[from] RattanOpError),
    #[error("Config error: {0}")]
    ConfigError(String),
    #[error("Channel error: {0}")]
    ChannelError(String),
    #[cfg(feature = "serde")]
    #[error("Serde error: {0}")]
    SerdeError(#[from] serde_json::Error),
}

#[derive(Debug, thiserror::Error)]
pub enum NsError {
    #[error("Can not create netns directory, {0}")]
    CreateNsDirError(std::io::Error),

    #[error("Can not create netns, {0}")]
    CreateNsError(std::io::Error),

    #[error("Can not open netns {0}, {1}")]
    OpenNsError(std::path::PathBuf, std::io::Error),

    #[error("Failed to close netns, {0}")]
    CloseNsError(nix::Error),

    #[error("Failed to mount {0}, {1}")]
    MountError(String, nix::Error),

    #[error("Failed to unmount {0}, {1}")]
    UnmountError(std::path::PathBuf, nix::Error),

    #[error("Failed to unshare, {0}")]
    UnshareError(nix::Error),

    #[error("Failed to join thread, {0}")]
    JoinThreadError(String),

    #[error("Can not setns, {0}")]
    SetNsError(nix::Error),
}

/// An error that may occur when parsing a MAC address string.
#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, thiserror::Error)]
pub enum MacParseError {
    /// Parsing of the MAC address contained an invalid digit.
    #[error("Invalid digit")]
    InvalidDigit,
    /// The MAC address did not have the correct length.
    #[error("Invalid length")]
    InvalidLength,
}

#[derive(Debug, thiserror::Error)]
pub enum VethError {
    #[error("Can not create veth pair, {0}")]
    CreateVethPairError(String),
    #[error("Encounter namespace error, {0}")]
    NsError(#[from] NsError),
    #[error("Encounter IO error, {0}")]
    IoError(#[from] std::io::Error),
    #[error("Encounter system error, {0}")]
    SystemError(#[from] nix::errno::Errno),
    #[error("Already in namespace {0}")]
    AlreadyInNamespace(String),
    #[error("Set Veth error, {0}")]
    SetError(String),
    #[error("Failed to build veth, {0}")]
    TokioRuntimeError(#[from] TokioRuntimeError),
}

#[derive(Debug, thiserror::Error)]
pub enum TokioRuntimeError {
    #[error("Failed to build runtime, {0}")]
    CreateError(#[from] std::io::Error),
}

#[derive(Debug, thiserror::Error)]
pub enum RattanCoreError {
    #[error("Failed to spawn rattan task, {0}")]
    SpawnError(String),
    #[error("Failed to add device, {0}")]
    AddDeviceError(String),
    #[error("Failed to send notify, {0}")]
    SendNotifyError(String),
    #[error("Unknown interface ID \"{0}\"")]
    UnknownIdError(String),
}

#[derive(Debug, thiserror::Error)]
pub enum RattanOpError {
    #[error("Failed to send operation {0:?} to rattan controller")]
    SendOpError(crate::control::RattanOp),
    #[error("Failed to receive operation result from rattan controller")]
    RecvOpResError,
    #[error("Operation result mismatch operation type, maybe implement error")]
    MismatchOpResError,
}

#[cfg(feature = "http")]
impl axum::response::IntoResponse for Error {
    fn into_response(self) -> axum::response::Response {
        use axum::http::StatusCode;
        use axum::response::Json;
        use serde_json::json;

        let status = match self {
            Error::MacParseError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::NsError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::VethError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::IoError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::MetalError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::TokioRuntimeError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            Error::RattanRadixError(_) => StatusCode::BAD_REQUEST,
            Error::RattanCoreError(RattanCoreError::UnknownIdError(_)) => StatusCode::NOT_FOUND,
            Error::RattanCoreError(_) => StatusCode::BAD_REQUEST,
            Error::RattanOpError(_) => StatusCode::BAD_REQUEST,
            Error::ConfigError(_) => StatusCode::BAD_REQUEST,
            Error::ChannelError(_) => StatusCode::INTERNAL_SERVER_ERROR,
            #[cfg(feature = "serde")]
            Error::SerdeError(_) => StatusCode::BAD_REQUEST,
        };

        (status, Json(json!({"error": self.to_string(),}))).into_response()
    }
}
