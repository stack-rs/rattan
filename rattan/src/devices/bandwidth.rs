use crate::devices::{Device, Packet};
use crate::error::Error;
use crate::metal::timer::Timer;
use crate::utils::sync::AtomicRawCell;
use async_trait::async_trait;
use netem_trace::{Bandwidth, Delay};
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};
use std::fmt::Debug;
use std::sync::Arc;
use tokio::sync::mpsc;
use tokio::time::Instant;
use tracing::{debug, info};

use super::{ControlInterface, Egress, Ingress};

const MAX_BANDWIDTH: Bandwidth = Bandwidth::from_bps(u64::MAX);

#[derive(Debug)]
struct BwPacket<P>
where
    P: Packet,
{
    ingress_time: Instant,
    packet: P,
}

pub struct BwDeviceIngress<P>
where
    P: Packet,
{
    ingress: mpsc::UnboundedSender<BwPacket<P>>,
}

impl<P> Clone for BwDeviceIngress<P>
where
    P: Packet,
{
    fn clone(&self) -> Self {
        Self {
            ingress: self.ingress.clone(),
        }
    }
}

impl<P> Ingress<P> for BwDeviceIngress<P>
where
    P: Packet + Send,
{
    fn enqueue(&self, packet: P) -> Result<(), Error> {
        // XXX(minhuw): handle possible error here
        self.ingress
            .send(BwPacket {
                ingress_time: Instant::now(),
                packet,
            })
            .unwrap();
        Ok(())
    }
}

// Requires the bandwidth to be less than 2^64 bps
fn transfer_time(length: usize, bandwidth: Bandwidth) -> Delay {
    let bits = length * 8;
    let capacity = bandwidth.as_bps() as u64;
    let seconds = bits as f64 / capacity as f64;
    Delay::from_secs_f64(seconds)
}

pub struct BwDeviceEgress<P>
where
    P: Packet,
{
    egress: mpsc::UnboundedReceiver<BwPacket<P>>,
    bandwidth: Arc<AtomicRawCell<Bandwidth>>,
    inner_bandwidth: Box<Bandwidth>,
    next_available: Instant,
    timer: Timer,
}

#[async_trait]
impl<P> Egress<P> for BwDeviceEgress<P>
where
    P: Packet + Send + Sync,
{
    async fn dequeue(&mut self) -> Option<P> {
        let packet = self.egress.recv().await.unwrap();
        if let Some(bandwidth) = self.bandwidth.swap_null() {
            self.inner_bandwidth = bandwidth;
            debug!(?self.inner_bandwidth, "Set inner bandwidth:");
        }
        let now = Instant::now();
        if packet.ingress_time >= self.next_available {
            // no need to wait, since the packet arrives after next_available
            let transfer_time = transfer_time(packet.packet.length(), *self.inner_bandwidth);
            self.next_available = packet.ingress_time + transfer_time;
        } else if now >= self.next_available {
            // the current time is already after next_available
            let transfer_time = transfer_time(packet.packet.length(), *self.inner_bandwidth);
            self.next_available += transfer_time;
        } else {
            // wait until next_available
            let wait_time = self.next_available - now;
            self.timer.sleep(wait_time).await.unwrap();
            let transfer_time = transfer_time(packet.packet.length(), *self.inner_bandwidth);
            self.next_available += transfer_time;
        }
        Some(packet.packet)
    }
}

#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
#[derive(Debug)]
pub struct BwDeviceConfig {
    bandwidth: Bandwidth,
}

impl BwDeviceConfig {
    pub fn new<T: Into<Bandwidth>>(bandwidth: T) -> Self {
        Self {
            bandwidth: bandwidth.into(),
        }
    }
}

pub struct BwDeviceControlInterface {
    bandwidth: Arc<AtomicRawCell<Bandwidth>>,
}

impl ControlInterface for BwDeviceControlInterface {
    type Config = BwDeviceConfig;

    fn set_config(&self, config: Self::Config) -> Result<(), Error> {
        if config.bandwidth > MAX_BANDWIDTH {
            return Err(Error::ConfigError(
                "Bandwidth should be less than 2^64 bps".to_string(),
            ));
        }
        info!("Setting bandwidth to: {:?}", config.bandwidth);
        self.bandwidth.store(Box::new(config.bandwidth));
        Ok(())
    }
}

pub struct BwDevice<P: Packet> {
    ingress: Arc<BwDeviceIngress<P>>,
    egress: BwDeviceEgress<P>,
    control_interface: Arc<BwDeviceControlInterface>,
}

impl<P> Device<P> for BwDevice<P>
where
    P: Packet + Send + Sync + 'static,
{
    type IngressType = BwDeviceIngress<P>;
    type EgressType = BwDeviceEgress<P>;
    type ControlInterfaceType = BwDeviceControlInterface;

    fn sender(&self) -> Arc<Self::IngressType> {
        self.ingress.clone()
    }

    fn receiver(&mut self) -> &mut Self::EgressType {
        &mut self.egress
    }

    fn into_receiver(self) -> Self::EgressType {
        self.egress
    }

    fn control_interface(&self) -> Arc<Self::ControlInterfaceType> {
        self.control_interface.clone()
    }
}

impl<P> BwDevice<P>
where
    P: Packet,
{
    pub fn new() -> BwDevice<P> {
        debug!("New BwDevice");
        let (rx, tx) = mpsc::unbounded_channel();
        let bandwidth = Arc::new(AtomicRawCell::new(Box::new(Bandwidth::from_bps(u64::MAX))));
        BwDevice {
            ingress: Arc::new(BwDeviceIngress { ingress: rx }),
            egress: BwDeviceEgress {
                egress: tx,
                bandwidth: Arc::clone(&bandwidth),
                inner_bandwidth: Box::new(Bandwidth::from_bps(u64::MAX)),
                next_available: Instant::now(),
                timer: Timer::new().unwrap(),
            },
            control_interface: Arc::new(BwDeviceControlInterface { bandwidth }),
        }
    }
}

impl<P> Default for BwDevice<P>
where
    P: Packet,
{
    fn default() -> Self {
        Self::new()
    }
}
