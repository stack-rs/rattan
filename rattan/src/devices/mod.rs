use async_trait::async_trait;
use etherparse::{Ethernet2Header, Ipv4Header};
#[cfg(feature = "serde")]
use serde::Deserialize;
use std::{fmt::Debug, sync::Arc};
use tokio::time::Instant;

use crate::error::Error;

pub mod bandwidth;
pub mod delay;
pub mod external;
pub mod loss;

pub trait Packet: Debug + 'static + Send {
    fn empty(maximum: usize) -> Self;
    fn from_raw_buffer(buf: &[u8]) -> Self;
    // Raw buffer length
    fn length(&self) -> usize;
    // Link layer length, i.e. the length of the Ethernet frame (not including the preamble, SFD, FCS and IPG)
    fn l2_length(&self) -> usize;
    // Network layer length
    fn l3_length(&self) -> usize;
    fn as_slice(&self) -> &[u8];
    fn as_raw_buffer(&mut self) -> &mut [u8];
    fn ether_hdr(&self) -> Option<Ethernet2Header>;
    fn ip_hdr(&self) -> Option<Ipv4Header>;

    fn get_timestamp(&self) -> Instant;
    fn set_timestamp(&mut self, timestamp: Instant);
}

#[derive(Debug)]
pub struct StdPacket {
    buf: Vec<u8>,
    timestamp: Instant,
}

impl Packet for StdPacket {
    fn empty(maximum: usize) -> Self {
        Self {
            buf: Vec::with_capacity(maximum),
            timestamp: Instant::now(),
        }
    }

    fn from_raw_buffer(buf: &[u8]) -> Self {
        Self {
            buf: buf.to_vec(),
            timestamp: Instant::now(),
        }
    }

    fn length(&self) -> usize {
        self.buf.len()
    }

    fn l2_length(&self) -> usize {
        self.buf.len()
    }

    fn l3_length(&self) -> usize {
        // 14 is the length of the Ethernet header
        self.buf.len() - 14
    }

    fn as_slice(&self) -> &[u8] {
        self.buf.as_slice()
    }

    fn as_raw_buffer(&mut self) -> &mut [u8] {
        self.buf.as_mut_slice()
    }

    fn ip_hdr(&self) -> Option<Ipv4Header> {
        if let Ok(result) = etherparse::Ethernet2Header::from_slice(self.buf.as_slice()) {
            if let Ok(ip_hdr) = etherparse::Ipv4Header::from_slice(result.1) {
                return Some(ip_hdr.0);
            }
        }
        None
    }

    fn ether_hdr(&self) -> Option<Ethernet2Header> {
        etherparse::Ethernet2Header::from_slice(self.buf.as_slice()).map_or(None, |x| Some(x.0))
    }

    fn get_timestamp(&self) -> Instant {
        self.timestamp
    }

    fn set_timestamp(&mut self, timestamp: Instant) {
        self.timestamp = timestamp;
    }
}

pub trait Ingress<P>: Send + Sync
where
    P: Packet,
{
    fn enqueue(&self, packet: P) -> Result<(), Error>;
}

#[async_trait]
pub trait Egress<P>: Send
where
    P: Packet,
{
    async fn dequeue(&mut self) -> Option<P>;
}

pub trait ControlInterface: Send + Sync + 'static {
    #[cfg(feature = "serde")]
    type Config: for<'a> Deserialize<'a> + Send;
    #[cfg(not(feature = "serde"))]
    type Config: Send;
    fn set_config(&self, config: Self::Config) -> Result<(), Error>;
}

#[async_trait]
pub trait Device<P>
where
    P: Packet,
{
    type IngressType: Ingress<P> + 'static;
    type EgressType: Egress<P> + 'static;
    type ControlInterfaceType: ControlInterface;

    fn sender(&self) -> Arc<Self::IngressType>;
    fn receiver(&mut self) -> &mut Self::EgressType;
    fn into_receiver(self) -> Self::EgressType;
    fn control_interface(&self) -> Arc<Self::ControlInterfaceType>;
}
