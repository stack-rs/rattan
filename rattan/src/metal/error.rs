#[cfg(feature = "camellia")]
use camellia::error::CamelliaError;

#[derive(Debug, thiserror::Error)]
pub enum MetalError {
    #[error("Encounter system error, {0}")]
    SystemError(#[from] nix::errno::Errno),
    #[error("Encounter IO error, {0}")]
    IoError(#[from] std::io::Error),
    #[cfg_attr(feature = "camellia", error("Encounter XDP error, {0}"))]
    #[cfg(feature = "camellia")]
    XDPError(#[from] CamelliaError),
    #[error("Encounter Rtnetlink error, {0}")]
    RtnetlinkError(#[from] rtnetlink::Error),
    #[error("not interested packet")]
    NotInterestedPacket,
}
