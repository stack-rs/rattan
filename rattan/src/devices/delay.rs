use crate::devices::{Device, Packet};
use crate::error::Error;
use crate::metal::timer::Timer;
use async_trait::async_trait;
use netem_trace::Delay;
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};
use std::fmt::Debug;
use std::sync::Arc;
use tokio::sync::mpsc;
use tokio::time::Instant;
use tracing::{debug, info};

use super::{ControlInterface, Egress, Ingress};

pub struct DelayDeviceIngress<P>
where
    P: Packet,
{
    ingress: mpsc::UnboundedSender<P>,
}

impl<P> Clone for DelayDeviceIngress<P>
where
    P: Packet,
{
    fn clone(&self) -> Self {
        Self {
            ingress: self.ingress.clone(),
        }
    }
}

impl<P> Ingress<P> for DelayDeviceIngress<P>
where
    P: Packet + Send,
{
    fn enqueue(&self, mut packet: P) -> Result<(), Error> {
        packet.set_timestamp(Instant::now());
        self.ingress
            .send(packet)
            .map_err(|_| Error::ChannelError("Data channel is closed.".to_string()))?;
        Ok(())
    }
}

pub struct DelayDeviceEgress<P>
where
    P: Packet,
{
    egress: mpsc::UnboundedReceiver<P>,
    delay: Delay,
    config_rx: mpsc::UnboundedReceiver<DelayDeviceConfig>,
    timer: Timer,
}

impl<P> DelayDeviceEgress<P>
where
    P: Packet + Send + Sync,
{
    fn set_config(&mut self, config: DelayDeviceConfig) {
        debug!(
            before = ?self.delay,
            after = ?config.delay,
            "Set inner delay:"
        );
        self.delay = config.delay;
    }
}

#[async_trait]
impl<P> Egress<P> for DelayDeviceEgress<P>
where
    P: Packet + Send + Sync,
{
    async fn dequeue(&mut self) -> Option<P> {
        let packet = match self.egress.recv().await {
            Some(packet) => packet,
            None => return None,
        };
        loop {
            tokio::select! {
                biased;
                Some(config) = self.config_rx.recv() => {
                    self.set_config(config);
                }
                _ = self.timer.sleep(packet.get_timestamp() + self.delay - Instant::now()) => {
                    break;
                }
            }
        }
        Some(packet)
    }
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Debug, Default, Clone)]
pub struct DelayDeviceConfig {
    #[cfg_attr(feature = "serde", serde(with = "humantime_serde"))]
    pub delay: Delay,
}

impl DelayDeviceConfig {
    pub fn new<T: Into<Delay>>(delay: T) -> Self {
        Self {
            delay: delay.into(),
        }
    }
}

pub struct DelayDeviceControlInterface {
    config_tx: mpsc::UnboundedSender<DelayDeviceConfig>,
}

impl ControlInterface for DelayDeviceControlInterface {
    type Config = DelayDeviceConfig;

    fn set_config(&self, config: Self::Config) -> Result<(), Error> {
        info!("Setting delay to {:?}", config.delay);
        self.config_tx
            .send(config)
            .map_err(|_| Error::ConfigError("Control channel is closed.".to_string()))?;
        Ok(())
    }
}

pub struct DelayDevice<P: Packet> {
    ingress: Arc<DelayDeviceIngress<P>>,
    egress: DelayDeviceEgress<P>,
    control_interface: Arc<DelayDeviceControlInterface>,
}

impl<P> Device<P> for DelayDevice<P>
where
    P: Packet + Send + Sync + 'static,
{
    type IngressType = DelayDeviceIngress<P>;
    type EgressType = DelayDeviceEgress<P>;
    type ControlInterfaceType = DelayDeviceControlInterface;

    fn sender(&self) -> Arc<Self::IngressType> {
        self.ingress.clone()
    }

    fn receiver(&mut self) -> &mut Self::EgressType {
        &mut self.egress
    }

    fn into_receiver(self) -> Self::EgressType {
        self.egress
    }

    fn control_interface(&self) -> Arc<Self::ControlInterfaceType> {
        Arc::clone(&self.control_interface)
    }
}

impl<P> DelayDevice<P>
where
    P: Packet,
{
    pub fn new<D: Into<Option<Delay>>>(delay: D) -> Result<DelayDevice<P>, Error> {
        let delay = delay.into().unwrap_or_default();
        debug!(?delay, "New DelayDevice");
        let (rx, tx) = mpsc::unbounded_channel();
        let (config_tx, config_rx) = mpsc::unbounded_channel();
        Ok(DelayDevice {
            ingress: Arc::new(DelayDeviceIngress { ingress: rx }),
            egress: DelayDeviceEgress {
                egress: tx,
                delay,
                config_rx,
                timer: Timer::new()?,
            },
            control_interface: Arc::new(DelayDeviceControlInterface { config_tx }),
        })
    }
}

#[cfg(test)]
mod tests {
    use std::time::{Duration, Instant};
    use tracing::{span, Level};

    use crate::devices::StdPacket;

    use super::*;

    // The tolerance of the accuracy of the delays, in ms
    const DELAY_ACCURACY_TOLERANCE: f64 = 1.0;
    // List of delay times to be tested
    const DELAY_TEST_TIME: [u64; 8] = [0, 2, 5, 10, 20, 50, 100, 500];

    #[test_log::test]
    fn test_delay_device() -> Result<(), Error> {
        let _span = span!(Level::INFO, "test_delay_device").entered();
        for testing_delay in DELAY_TEST_TIME {
            let rt = tokio::runtime::Runtime::new()?;

            let _guard = rt.enter();

            info!("Creating device with {}ms delay", testing_delay);
            let device_config = DelayDeviceConfig::new(Duration::from_millis(testing_delay));
            let builder = device_config.into_factory::<StdPacket>();
            let device = builder(rt.handle())?;
            let ingress = device.sender();
            let mut egress = device.into_receiver();

            info!("Testing delay time for {}ms delay device", testing_delay);
            let mut delays: Vec<f64> = Vec::new();

            for _ in 0..10 {
                let test_packet = StdPacket::from_raw_buffer(&[0; 256]);
                let start = Instant::now();
                ingress.enqueue(test_packet)?;
                let received = rt.block_on(async { egress.dequeue().await });

                // Use mircro second to get precision up to 0.001ms
                let duration = start.elapsed().as_micros() as f64 / 1000.0;

                delays.push(duration);

                // Should never loss packet
                assert!(received.is_some());

                let received = received.unwrap();

                // The length should be correct
                assert!(received.length() == 256);
            }

            info!(
                "Tested delays for {}ms delay device: {:?}",
                testing_delay, delays
            );

            let average_delay = delays.iter().sum::<f64>() / 10.0;
            debug!("Delays: {:?}", delays);
            info!(
                "Average delay: {:.3}ms, error {:.1}ms",
                average_delay,
                (average_delay - testing_delay as f64).abs()
            );
            // Check the delay time
            assert!((average_delay - testing_delay as f64) <= DELAY_ACCURACY_TOLERANCE);
        }

        Ok(())
    }

    #[test_log::test]
    fn test_delay_device_config_update() -> Result<(), Error> {
        let _span = span!(Level::INFO, "test_delay_device_config_update").entered();
        let rt = tokio::runtime::Runtime::new()?;

        let _guard = rt.enter();

        info!("Creating device with 10ms delay");
        let device_config = DelayDeviceConfig::new(Duration::from_millis(10));
        let builder = device_config.into_factory::<StdPacket>();
        let device = builder(rt.handle())?;
        let config_changer = device.control_interface();
        let ingress = device.sender();
        let mut egress = device.into_receiver();

        //Test whether the packet will wait longer if the config is updated
        let test_packet = StdPacket::from_raw_buffer(&[0; 256]);

        let start = Instant::now();
        ingress.enqueue(test_packet)?;

        // Wait for 5ms, then change the config to let the delay be longer
        std::thread::sleep(Duration::from_millis(5));
        config_changer.set_config(DelayDeviceConfig::new(Duration::from_millis(20)))?;

        let received = rt.block_on(async { egress.dequeue().await });

        let duration = start.elapsed().as_micros() as f64 / 1000.0;

        info!("Delay after update: {}ms", duration);

        assert!(received.is_some());
        let received = received.unwrap();
        assert!(received.length() == 256);

        assert!((duration - 20.0).abs() <= DELAY_ACCURACY_TOLERANCE);

        // Test whether the packet will be returned immediately when the new delay is less than the already passed time
        let test_packet = StdPacket::from_raw_buffer(&[0; 256]);

        let start = Instant::now();
        ingress.enqueue(test_packet)?;

        // Wait for 15ms, then change the config back to 10ms
        std::thread::sleep(Duration::from_millis(15));
        config_changer.set_config(DelayDeviceConfig::new(Duration::from_millis(10)))?;

        let received = rt.block_on(async { egress.dequeue().await });

        let duration = start.elapsed().as_micros() as f64 / 1000.0;

        info!("Delay after update: {}ms", duration);

        assert!(received.is_some());
        let received = received.unwrap();
        assert!(received.length() == 256);

        assert!((duration - 15.0).abs() <= DELAY_ACCURACY_TOLERANCE);

        Ok(())
    }
}
