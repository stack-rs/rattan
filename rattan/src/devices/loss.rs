use crate::devices::{Device, Packet};
use crate::error::Error;
use crate::utils::sync::AtomicRawCell;
use async_trait::async_trait;
use netem_trace::LossPattern;
use rand::Rng;
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};
use std::fmt::Debug;
use std::sync::Arc;
use tokio::sync::mpsc;
use tracing::{debug, info};

use super::{ControlInterface, Egress, Ingress};

#[derive(Debug)]
struct LossPacket<P>
where
    P: Packet,
{
    packet: P,
}

pub struct LossDeviceIngress<P>
where
    P: Packet,
{
    ingress: mpsc::UnboundedSender<LossPacket<P>>,
}

impl<P> Clone for LossDeviceIngress<P>
where
    P: Packet,
{
    fn clone(&self) -> Self {
        Self {
            ingress: self.ingress.clone(),
        }
    }
}

impl<P> Ingress<P> for LossDeviceIngress<P>
where
    P: Packet + Send,
{
    fn enqueue(&self, packet: P) -> Result<(), Error> {
        self.ingress
            .send(LossPacket { packet })
            .map_err(|_| Error::ChannelError("Data channel is closed.".to_string()))?;
        Ok(())
    }
}

pub struct LossDeviceEgress<P, R>
where
    P: Packet,
    R: Rng,
{
    egress: mpsc::UnboundedReceiver<LossPacket<P>>,
    pattern: Arc<AtomicRawCell<LossPattern>>,
    inner_pattern: Box<LossPattern>,
    /// How many packets have been lost consecutively
    prev_loss: usize,
    rng: R,
}

#[async_trait]
impl<P, R> Egress<P> for LossDeviceEgress<P, R>
where
    P: Packet + Send + Sync,
    R: Rng + Send + Sync,
{
    async fn dequeue(&mut self) -> Option<P> {
        let packet = self.egress.recv().await.unwrap();
        if let Some(pattern) = self.pattern.swap_null() {
            self.inner_pattern = pattern;
            debug!(?self.inner_pattern, "Set inner pattern:");
        }
        let loss_rate = match self.inner_pattern.get(self.prev_loss) {
            Some(&loss_rate) => loss_rate,
            None => *self.inner_pattern.last().unwrap_or(&0.0),
        };
        let rand_num = self.rng.gen_range(0.0..1.0);
        if rand_num < loss_rate {
            self.prev_loss += 1;
            None
        } else {
            self.prev_loss = 0;
            Some(packet.packet)
        }
    }
}

// Loss pattern will repeat the last value until stop dropping packets.
// For example, the pattern [0.1, 0.2, 0.3] means [0.1, 0.2, 0.3, 0.3, 0.3, ...].
// Set the last value of the pattern to 0 to limit the maximum number of consecutive packet losses.
// If you want to drop packets i.i.d., just set the pattern to a single number, such as [0.1].
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
#[derive(Debug)]
pub struct LossDeviceConfig {
    pattern: LossPattern,
}

impl LossDeviceConfig {
    pub fn new<T: Into<LossPattern>>(pattern: T) -> Self {
        Self {
            pattern: pattern.into(),
        }
    }
}

pub struct LossDeviceControlInterface {
    /// Stored as nanoseconds
    pattern: Arc<AtomicRawCell<LossPattern>>,
}

impl ControlInterface for LossDeviceControlInterface {
    type Config = LossDeviceConfig;

    fn set_config(&self, config: Self::Config) -> Result<(), Error> {
        info!("Set loss pattern to: {:?}", config.pattern);
        self.pattern.store(Box::new(config.pattern));
        Ok(())
    }
}

pub struct LossDevice<P: Packet, R: Rng> {
    ingress: Arc<LossDeviceIngress<P>>,
    egress: LossDeviceEgress<P, R>,
    control_interface: Arc<LossDeviceControlInterface>,
}

impl<P, R> Device<P> for LossDevice<P, R>
where
    P: Packet + Send + Sync + 'static,
    R: Rng + Send + Sync + 'static,
{
    type IngressType = LossDeviceIngress<P>;
    type EgressType = LossDeviceEgress<P, R>;
    type ControlInterfaceType = LossDeviceControlInterface;

    fn sender(&self) -> Arc<Self::IngressType> {
        self.ingress.clone()
    }

    fn receiver(&mut self) -> &mut Self::EgressType {
        &mut self.egress
    }

    fn into_receiver(self) -> Self::EgressType {
        self.egress
    }

    fn control_interface(&self) -> Arc<Self::ControlInterfaceType> {
        Arc::clone(&self.control_interface)
    }
}

impl<P, R> LossDevice<P, R>
where
    P: Packet,
    R: Rng,
{
    pub fn new(rng: R) -> LossDevice<P, R> {
        debug!("New LossDevice");
        let (rx, tx) = mpsc::unbounded_channel();
        let pattern = Arc::new(AtomicRawCell::new(Box::default()));
        LossDevice {
            ingress: Arc::new(LossDeviceIngress { ingress: rx }),
            egress: LossDeviceEgress {
                egress: tx,
                pattern: Arc::clone(&pattern),
                inner_pattern: Box::default(),
                prev_loss: 0,
                rng,
            },
            control_interface: Arc::new(LossDeviceControlInterface { pattern }),
        }
    }
}
