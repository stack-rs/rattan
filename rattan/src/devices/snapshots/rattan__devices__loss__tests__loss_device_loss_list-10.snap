---
source: rattan/src/devices/loss.rs
expression: loss_seq
---
[
  true,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  true,
  true,
  false,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  false,
  false,
  true,
  true,
  false,
  true,
  true,
  true,
  false,
  true,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  true,
  false,
  true,
  true,
  false,
  true,
  false,
  true,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  true,
  false,
  true,
  true,
  false,
  true,
  true,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  true,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  true,
  true,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  true,
  false,
  false,
  true,
  false,
  true,
  false,
  true,
  false,
  true,
  false
]
